core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +842,186 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
+			continue;
+
+		memcpy(&dump_ctx[n].uuid, &ctx->ts_ctx.uuid,
+		       sizeof(ctx->ts_ctx.uuid));
+		dump_ctx[n].panicked = ctx->panicked;
+		dump_ctx[n].is_user_ta = is_user_ta_ctx(&ctx->ts_ctx);
+		TAILQ_FOREACH(sess, open_sessions, link) {
+			if (sess->ts_sess.ctx == &ctx->ts_ctx) {
+				if (cnt == MAX_DUMP_SESS_NUM)
+					break;
+
+				dump_ctx[n].sess_id[cnt] = sess->id;
+				cnt++;
+			}
+		}
+
+		dump_ctx[n].sess_num = cnt;
+		n++;
+	}
+}
+
+static TEE_Result dump_ta_stats(struct tee_ta_dump_ctx *dump_ctx,
+				struct tee_ta_dump_stats *dump_stats,
+				size_t ta_count)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_param param = { };
+	unsigned int i = 0;
+	unsigned int j = 0;
+
+	nsec_sessions_list_head(&open_sessions);
+
+	for (i = 0; i < ta_count; i++) {
+		struct tee_ta_dump_stats *stats = &dump_stats[i];
+
+		memcpy(&stats->uuid, &dump_ctx[i].uuid,
+		       sizeof(dump_ctx[i].uuid));
+		stats->panicked = dump_ctx[i].panicked;
+		stats->sess_num = dump_ctx[i].sess_num;
+
+		/* find a session from dump context */
+		sess = NULL;
+		for (j = 0; j < dump_ctx[i].sess_num; j++) {
+			uint32_t sess_id = dump_ctx[i].sess_id[j];
+
+			sess = tee_ta_get_session(sess_id,
+						  true,
+						  open_sessions);
+			if (sess)
+				break;
+		}
+
+		if (!sess)
+			continue;
+		/* if session is existing, get its heap stats */
+		memset(&param, 0, sizeof(struct tee_ta_param));
+		param.types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+					      TEE_PARAM_TYPE_VALUE_OUTPUT,
+					      TEE_PARAM_TYPE_VALUE_OUTPUT,
+					      TEE_PARAM_TYPE_NONE);
+		res = dump_ta_memstats(sess, TEE_TIMEOUT_INFINITE, &param);
+		if (res == TEE_SUCCESS) {
+			stats->heap.allocated = param.u[0].val.a;
+			stats->heap.max_allocated = param.u[0].val.b;
+			stats->heap.size = param.u[1].val.a;
+			stats->heap.num_alloc_fail = param.u[1].val.b;
+			stats->heap.biggest_alloc_fail = param.u[2].val.a;
+			stats->heap.biggest_alloc_fail_used = param.u[2].val.b;
+		}
+		tee_ta_put_session(sess);
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_ta_instance_stats(void *buf, uint32_t *buf_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_dump_stats *dump_stats = NULL;
+	struct tee_ta_dump_ctx *dump_ctx = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	size_t sz = 0;
+	size_t ta_count = 0;
+
+	if (!buf_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&tee_ta_mutex);
+
+	/*
+	 * Go through all available TA and calc out the actual buffer size.
+	 */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link)
+		if (is_user_ta_ctx(&ctx->ts_ctx))
+			ta_count++;
+
+	sz = sizeof(struct tee_ta_dump_stats) * ta_count;
+	if (!buf || *buf_size < sz) {
+		/*
+		 * buf is null or pass size less than actual size
+		 * means caller try to query the buffer size.
+		 * update *buf_size.
+		 */
+		*buf_size = sz;
+		res = TEE_ERROR_SHORT_BUFFER;
+	} else {
+		dump_stats = (struct tee_ta_dump_stats *)buf;
```

Comment:
This asumes `buf` is 32bit aligned which it may not.
```patch
+	} else if (!IS_ALIGNED_WITH_TYPE(buf, uint32_t)) {
+		DMSG("Data aligment");
+		res = TEE_ERROR_BAD_PARAMETERS;
 	} else {
 		dump_stats = (struct tee_ta_dump_stats *)buf;
```

Reply:
good catch!